keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt

```diff
@@ -156,25 +157,53 @@ class CombinedRepository(
   }
 
   /**
-   * Validates that resources have unique ids, throws an exception if invalid
+   * Run validation checks against delivery config to ensure:
+   *
+   * - resources have unique ids
+   * - artifacts have unique references
+   *
+   * Throws an exception if config fails any checks
    */
   private fun validate(config: DeliveryConfig) {
+
+    // helper function to get duplicates in a list
+    val duplicates: (List<String>) -> List<String> =
+      { ls -> ls.groupingBy { it }.eachCount().filter { it.value > 1 }.keys.toList() }
+
+    /**
+     * check: resources have unique ids
+     */
+
     val resources = config.environments.map { it.resources }.flatten().map { it.id }
-    val distinct = resources.distinct()
+    val duplicateResources = duplicates(resources)
 
-    if (resources.size != distinct.size) {
-      val duplicates = resources.groupingBy { it }.eachCount().filter { it.value > 1 }.keys.toList()
+    if (duplicateResources.isNotEmpty()) {
       val envToResources: Map<String, MutableList<String>> = config.environments
         .map { env -> env.name to env.resources.map { it.id }.toMutableList() }.toMap()
       val envsAndDuplicateResources = envToResources
         .filterValues { rs: MutableList<String> ->
           // remove all the resources we don't care about from this mapping
-          rs.removeIf { it !in duplicates }
+          rs.removeIf { it !in duplicateResources }
           // if there are resources left that we care about, leave it in the map
           rs.isNotEmpty()
         }
-      log.error("Validation failed for ${config.name}, duplicates found: $envsAndDuplicateResources")
-      throw DuplicateResourceIdException(duplicates, envsAndDuplicateResources)
+      log.error("Validation failed for ${config.name}, duplicates resource ids found: $envsAndDuplicateResources")
+      throw DuplicateResourceIdException(duplicateResources, envsAndDuplicateResources)
+    }
+
+    /**
+     * check: artifacts have unique references
+     */
+    val refs = config.artifacts.map { it.reference }
+    val duplicateRefs = duplicates(refs)
+
+    if (duplicateRefs.isNotEmpty()) {
+      val duplicatesArtifactNameToRef: Map<String, String> = config.artifacts
+        .filter { duplicateRefs.contains(it.reference) }
+        .map { art -> art.name to art.reference }.toMap()
```

Comment:
Instead of combining `.map` and `toMap()`, you can use `.associate { art -> art.name to art.reference }`

Reply:
TIL!

Reply:
Addressed in 1bea51c.